File: 0049-group-anagrams/0049-group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        //        edge case
        if (strs.length == 0) return new ArrayList();
//        declare a hashmap
        Map<String, List> hash = new HashMap<String, List>();
//        loop through strings
        for (String s : strs) {
//            conver string to array of characters
            char[] ca = s.toCharArray();
//            sort the characters array
            Arrays.sort(ca);
//            convert sorted array to string
            String key = String.valueOf(ca);
//            if key is not in the hashmap add it
            if (!hash.containsKey(key)) hash.put(key, new ArrayList());
//            add the string to the list of the key
            hash.get(key).add(s);
        }
//        return only the vals of hashmap
        return new ArrayList(hash.values());
    }
}

File: 0049-group-anagrams/0049-group-anagrams.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        #create a dictionary
        dict = {}
        #loop through the list
        for i in strs:
            #sort the word:
            sorted_word = ''.join(sorted(i))
            # if the sorted word is not in the dictionary, add it as key with the original string as value
            if sorted_word not in dict:
                dict[sorted_word] = [i]
            # if the sorted word is in the dictionary, add the word as a value to the existing key
            else:
                dict[sorted_word].append(i)
            # convert the dictionary to a list of lists and return its values
        return list(dict.values())   